### ISSUE TITLE
[Fluid] 101 small API extensions

### ISSUE THREAD

**Issue by DaanVanYperen** [junkdog/artemis-odb#461]
Undeveloped extension ideas for #458
- [ ] **Versatile type exposure**
  - [ ] Single value type automatic defaullt getter/setter
  - [ ] Built in smart type handling (Pos2, Pos3)
    - [ ] But let users disable!
  - [x] Extendable type handling.
  - [ ] Better fluid interface for arrays (`m(p1) m(p1,p2) m(p1,p2,p3), m(...p)`.)
  - [ ] Fluid interface for lists (add/remove)
  - [ ] Fluid interface for `@EntityId` annotated fields `class Using { @EntityId int using; } ` `.using(item)`
- [ ] **Mirror/copy**
  - [ ] Component value copy
  - [ ] Entity mirror
  - [ ] init entity from archetype
  - [ ] `E("prefab/player.json").pos(5,5);`
- [ ] **Misc**
  - [ ] E.toggleAnchor();
  - [ ] E::remove/add by type. 
  - [ ] json convenience
  - [ ] Annotation to expose properties as (part of) the default parameters without having to define a manual set method.
- [ ] **Resolving**
  - [x] Fetch tags. `E.withTag("tag").pos(5,5)`
  - [x] Fetch sets `E.withGroup("group")`
- [ ] **Plugin**
  - [ ] Hook in third party generators in maven/gradle plugin (`IterativeModelStrategy`/`BuilderModelStrategy`)
- [ ] **Advanced**
  - [ ] Fluid Interface for different scopes (library and game).
  - [ ] Generate for multithread support.
  - [ ] Prevent loss of component method parameter names in fluid interface.
  - [ ] Support generics

**Finished**
- [x] **Annotations**
  - [x] Explicitly Exclude components, methods from fluid interface.
  - [x] Force contract for fluid interface methods.
- [x] Misc
  - [X] tag, group convenience
  - [x] Write tests that actually test the E interface methods.
  - [x] `E(id).deleteFromWorld()`
  - [x]  Disable default generators. (`FieldComponentBooleanAccessorStrategy`)
  - [x] Fluid entity systems.

## Advanced API

**Value copy**
Shallow copy of values from source to target.

```
e = E(sourceId);
E(targetId).pos(e).angle(e);
```

**Init from entity**
Create all components from source into target, and copy all values. basically a clone for empty entities, 
mixin for filled entities.

```
e = E(sourceId);
E(targetId).mirror(e);
```

**Init from archetype**

```
E(archetype).pos(e).angle(e);
```

**User convenience methods?!**

Basically some convenience to expose your own convenience methods.

``` java
E(id).myMethod(a).myMethod(b);
```

**Json convenience**
Can't really gauge utility since single entity is a bit of an awkward scope.

```
String entityJson = e.json();
e2.json(entityJson); 
String componentJson = e.posJson();
e2.posJson(e2);
```
# Fluid Entity Systems

Just because I'm particularly lazy we could even consider entity system tailored to hide away all the E(id) stuff.

``` java
class MySystem extends FluidIteratingSystem<E> {

  public void process(E e)
  {
       e.pos(5,5).script(sequence(..));
  }

}
```

Would be doable to provide adapters for entity subscriptions as well.

**Comment by DaanVanYperen** [comment]
Work is done, just needs a merge. closing ticket.

**Comment by DaanVanYperen** [comment]
Oops wrong ticket D:

**Comment by gjroelofs** [comment]
Got me all excited for a second!

**Comment by DaanVanYperen** [comment]
Bulk spring cleaning. Ticket too big.
